File: sdk/ts/coin/src/customers.ts
import type { CoinClient } from "./client.js";
import type { CustomerBalance, ListBooksResponse, ListBookCreditChangeLogsResponse } from "./types.js";

export class CustomersAPI {
  /** @internal */
  private readonly client: CoinClient;

  /** @internal */
  constructor(client: CoinClient) {
    this.client = client;
  }

  async getBalance(outUserID: string): Promise<CustomerBalance> {
    outUserID = (outUserID ?? "").trim();
    if (!outUserID) throw new Error("out_user_id is required");
    return this.client._do<CustomerBalance>(
      "GET",
      "/api/v1/customers/balance",
      { out_user_id: outUserID },
    );
  }

  async listBooks(outUserID: string): Promise<ListBooksResponse> {
    outUserID = (outUserID ?? "").trim();
    if (!outUserID) throw new Error("out_user_id is required");
    return this.client._do<ListBooksResponse>(
      "GET",
      "/api/v1/customers/books",
      { out_user_id: outUserID },
    );
  }

  async listBookCreditChangeLogs(bookNo: string): Promise<ListBookCreditChangeLogsResponse> {
    bookNo = (bookNo ?? "").trim();
    if (!bookNo) throw new Error("book_no is required");
    return this.client._do<ListBookCreditChangeLogsResponse>(
      "GET",
      `/api/v1/customers/books/${bookNo}/change-logs`,
    );
  }
}

File: sdk/go/coin/customers.go
package coin

import (
	"context"
	"fmt"
	"net/url"
	"strings"
)

// CustomersAPI contains customer-related calls for merchant apps.
type CustomersAPI struct {
	client *Client
}

// CustomerBalance represents the balance response for a customer.
type CustomerBalance struct {
	OutUserID         string `json:"out_user_id"`
	AccountNo         string `json:"account_no"`
	Balance           int64  `json:"balance"`
	AvailableBalance  int64  `json:"available_balance"`
	BookEnabled       bool   `json:"book_enabled"`
}

// GetBalance retrieves the balance for a customer identified by out_user_id.
func (c *CustomersAPI) GetBalance(ctx context.Context, outUserID string) (CustomerBalance, error) {
	outUserID = strings.TrimSpace(outUserID)
	if outUserID == "" {
		return CustomerBalance{}, fmt.Errorf("out_user_id is required")
	}
	q := url.Values{}
	q.Set("out_user_id", outUserID)
	var out CustomerBalance
	if err := c.client.do(ctx, "GET", "/api/v1/customers/balance", q, nil, &out); err != nil {
		return CustomerBalance{}, err
	}
	return out, nil
}
